### PR TITLE
JOSS paper

### PR DESCRIPTION
**Pull request by mivanit**
note that although a `paper.pdf` is in the repo, the canonical version is that which is compiled with the github action. You can see the most up-to-date version under the latest action here, under "artifacts" at the bottom: https://github.com/understanding-search/maze-dataset/actions/workflows/build-paper.yml

for example: https://github.com/understanding-search/maze-dataset/actions/runs/14063585067/artifacts/2816813853

### CODE REVIEW

File: docs/paper/paper.md [hunk 1]
```diff
@@ -18,6 +18,8 @@ authors:
   - name: Rusheb Shah
     orcid: 0009-0008-3058-0217
   - name: Brandon Knutson
+  - orcid: 0009-0004-8413-0239
+    affiliation: 1
   - name: Cecilia Diniz Behn
     affiliation: 1
```
Comment (cdinizbehn) [inline docs/paper/paper.md:24]: add orchid: 0000-0002-8078-5105

File: docs/paper/paper.tex [hunk 2]
```diff
@@ -0,0 +1,1258 @@
...
+
+\begin{figure} 
+  \centering
+  \begin{minipage}{5in}
+    \footnotesize
+    \input{figures/outputs-tokens-colored.tex}
+  \end{minipage}
+  \caption{
+    Example text output format with token regions highlighted.
+    \colorbox[RGB]{ 217,210,233 }{Adjacency list}: text representation of the lattice graph,
```
Comment (aaron-sandoval) [inline docs/paper/paper.tex:781]: lattice subgraph?

File: maze_dataset/tokenization/__init__.py [hunk 3]
```diff
@@ -2,7 +2,74 @@
 
 - `MazeTokenizerModular` is the new recommended way to do this as of 1.0.0
 - legacy `TokenizationMode` enum and `MazeTokenizer` class for supporting existing code
-- a whole lot of helper classes and functions
+- a variety of helper classes and functions
+
+There are many algorithms by which one might tokenize a 2D maze into a 1D format usable by autoregressive text models. Training multiple models on the encodings output from each of these algorithms may produce very different internal representations, learned solution algorithms, and levels of performance. To explore how different maze tokenization algorithms affect these models, the `MazeTokenizerModular` class contains a rich set of options to customize how mazes are stringified. This class contains 19 discrete parameters, resulting in 5.9 million unique tokenizers. But wait, there's more! There are 6 additional parameters available in the library which are untested but further expand the the number of tokenizers by a factor of $44/3$ to 86 million.
```
Comment (aaron-sandoval) [inline maze_dataset/tokenization/__init__.py:7]: The paper looks so nice and professional now I think we should remove "But wait, there's more!" :(

File: docs/paper/paper.md [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+
+
+## Related Works
+
+A multitude of public and open-source software packages exist for generating mazes [@easy_to_hard; @gh_Ehsan_2022; @gh_Nemeth_2019]. However, nearly all of these packages generate and store mazes in a form that is not optimized for storage space or, more importantly, computer readability. The mazes produces by other packages are usually rasterized or in some form of image, rather than the underlying graph structure, and this makes it difficult to work with these datasets.
```
Comment (traeuker) [inline docs/paper/paper.md:90]: computer readability --> "in a form that's more accessible to computational reasoning" or something similar?

Comment (traeuker) [reply docs/paper/paper.md:90]: "The mazes produces by other packages are usually rasterized or in **some form of image**" --> I would say:
"The mazes produced by other packages are typically represented as rasterized images or other visual formats."

File: docs/paper/paper.md [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+
+- Most prior works provide mazes in some kind of image or raster format, and we provide a variety of similar output formats:
```
Comment (traeuker) [inline docs/paper/paper.md:92]: "some kind of image" see above

File: docs/paper/paper.md [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+
+There are many algorithms by which one might tokenize a 2D maze into a 1D format usable by autoregressive text models. Training multiple models on the encodings output from each of these algorithms may produce very different internal representations, learned solution algorithms, and levels of performance. To allow exploration of how different maze tokenization algorithms affect these models, the [`MazeTokenizerModular`](https://understanding-search.github.io/maze-dataset/maze_dataset/tokenization.html#MazeTokenizerModular) class contains a rich set of options to customize how mazes are stringified. This class contains 19 discrete parameters, resulting in over 5.8 million unique tokenizers. But wait, there's more! There are 6 additional parameters available in the library which are untested but further expand the the number of tokenizers by a factor of $44/3$ to 86 million.
```
Comment (traeuker) [inline docs/paper/paper.md:183]: "levels of performance." --> probably cite our paper? 

"But wait, there's more!" would remove but is great for a poster

Comment (traeuker) [reply docs/paper/paper.md:183]: "which are untested" -> Elaborate what this means!

File: docs/paper/paper.md [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+Planned improvements to the `maze-dataset` include adding more generation algorithms (such as Prim's algorithm [@jarnik-prim; @prim; @dijkstra-prim] and Kruskal's algorithm [@kruskal], among others [@mazegen_analysis]), adding the ability to augment a maze with an adjacency list to add "shortcuts" to the maze, and resolving certain limitations detailed in Section [Limitations](#limitations). -->
+
+# Acknowledgements
+
+This work was partially supported by and many of the authors were brought together by AI Safety Camp and AI Safety Support. We would like to thank our former collaborators at AI Safety Camp and other users and contributors to the  `maze-dataset` package: Naveen Arunachalam, Benji Berczi, Guillaume Corlouer, William Edwards, Leon Eshuijs, Chris Mathwin, Lucia Quirke, Can Rager, Adrians Skapars, Johannes Treutlein, and Dan Valentine.
+
+
+This work was partially funded by National Science Foundation awards DMS-2110745 and DMS-2309810. We are also grateful to LTFF and FAR Labs for hosting three of the authors for a Residency Visit, and to various members of FAR’s technical staff for their advice.
```
Comment (traeuker) [inline docs/paper/paper.md:330]: "for hosting MII, AFS and TR"

File: docs/paper/paper.md [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+
+For interpretability and behavioral research, algorithmic tasks offer benefits by allowing systematic data generation and task decomposition, as well as simplifying the process of circuit discovery [@interpretability-survery]. Although mazes are well suited for these investigations, we have found that existing maze generation packages [@cobbe2019procgen; @harriesMazeExplorerCustomisable3D2019; @gh_Ehsan_2022; @gh_Nemeth_2019; @easy_to_hard] do not support flexible maze generation algorithms that provide fine-grained control of generation parameters and the ability to easily transform between multiple representations of the mazes (Raster, Textual, Tokenized) for training and testing models.
```
Comment (afspies) [inline docs/paper/paper.md:85]: we have found -> we found

Comment (afspies) [inline docs/paper/paper.md:85]: lack support both for fine-grained and flexible maze generation, as well as transforming between multiple representations

File: docs/paper/paper.md [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+  - [`MazePlot`](https://understanding-search.github.io/maze-dataset/maze_dataset/plotting.html#MazePlot) provides a feature-rich plotting utility with support for multiple paths, heatmaps over positions, and more. This is similar to the outputs of [@mdl-suite; @mathematica-maze; @mazegenerator-net; @gh_Ehsan_2022]
+
+
+- The text format provided by [`SolvedMaze(...).as_tokens()`](https://understanding-search.github.io/maze-dataset/maze_dataset.html#MazeDataset.as_tokens) is similar to that of [@eval-LLM-graphs], but provides over 5.8 million unique formats for converting mazes to a text stream, detailed in \autoref{tokenized-output-formats}.
+
+- For rigorous investigations of the response of a model to various distributional shifts, preserving metadata about the generation algorithm with the dataset itself is essential. To this end, our package efficiently stores the dataset along with its metadata in a single human-readable file [@zanj]. As far as we are aware, no existing packages do this reliably.
+
+- Storing mazes as images is not only difficult to work with, but also inefficient. We use an extremely efficient method detailed in \autoref{implementation}.
```
Comment (afspies) [inline docs/paper/paper.md:102]: highly efficient

File: docs/paper/paper.md [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+)
+dataset: MazeDataset = MazeDataset.from_config(cfg)
+```
+
+When initializing mazes, further configuration options can be specified through the [`from_config()`](https://understanding-search.github.io/maze-dataset/maze_dataset.html#MazeDataset.from_config) factory method as necessary. Options allow for saving/loading existing datasets instead of regenerating, and parallelization options for generation. Available maze generation algorithms are static methods of the [`LatticeMazeGenerators`](https://understanding-search.github.io/maze-dataset/maze_dataset.html#LatticeMazeGenerators) class and include generation algorithms based on randomized depth-first search, Wilson's algorithm [@wilson], percolation [@percolation; @percolation-clustersize], Kruskal's algorithm [@kruskal1956shortest], and others.
+
+Furthermore, a dataset of mazes can be filtered to satisfy certain properties. Custom filters can be specified, and some filters are included in [`MazeDatasetFilters`](https://understanding-search.github.io/maze-dataset/maze_dataset/dataset/filters.html#MazeDatasetFilters). For example, we can require a minimum path length from origin to target of 3:
```
Comment (afspies) [inline docs/paper/paper.md:128]: minimum path length of three steps from the origin to the target

File: docs/paper/paper.md [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+
+```python
+dataset_filtered: MazeDataset = dataset.filter_by.path_length(min_length=3)
+```
+
+All implemented maze generation algorithms are stochastic by nature. For reproducibility, the `seed` parameter of [`MazeDatasetConfig`](https://understanding-search.github.io/maze-dataset/maze_dataset.html#MazeDatasetConfig) may be set. In practice, we do not find that exact duplicates of mazes are generated with any meaningful frequency, even when generating large datasets.
```
Comment (afspies) [inline docs/paper/paper.md:134]: we find that duplicate mazes are generated very infrequently, ..

File: docs/paper/paper.md [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+\end{figure}
+
+In previous work, maze tasks have been used with Recurrent Convolutional Neural Network (RCNN) derived architectures [@deepthinking]. To facilitate the use of our package in this context, we replicate the format of [@easy_to_hard] and provide the [`RasterizedMazeDataset`](https://understanding-search.github.io/maze-dataset/maze_dataset/dataset/rasterized.html#RasterizedMazeDataset) class which returns rasterized pairs of (input, target) mazes as shown in \autoref{fig:e2h-raster} below.
+
+![Input is the rasterized maze without the path marked (left), and provide as a target the maze with all but the correct path removed. Configuration options exist to adjust whether endpoints are included and if empty cells should be filled in.](figures/maze-raster-input-target.pdf){#fig:e2h-raster width=30%}
+
+
+## Tokenized Output Formats {#tokenized-output-formats}
+
+Autoregressive transformer models can be quite sensitive to the exact format of the input data, and may use delimiter tokens to perform reasoning steps [@pfau2024dotbydot]. To facilitate systematic investigation of this, we provide a variety of tokenized output formats.
```
Comment (afspies) [inline docs/paper/paper.md:179]: the dot by dot reference seems under-explained and unnecessary here

File: docs/paper/paper.md [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+
+Parallelization is implemented via the `multiprocessing` module in the Python standard library, and parallel generation can be controlled via keyword arguments to [`MazeDataset.from_config()`](https://understanding-search.github.io/maze-dataset/maze_dataset.html#MazeDataset.from_config).
+
+# Usage in Research
+
+This package was originally built for the needs of the [@maze-transformer-github] project, which aims to investigate spatial planning and world models in autoregressive transformer models trained on mazes [@ivanitskiy2023structuredworldreps; @spies2024causalworldmodels]. This project has also adapted itself to be useful for work on understanding the mechanisms by which recurrent convolutional and implicit networks [@fung2022jfb] solve mazes given a rasterized view [@knutson2024logicalextrapolation], and for this we match the output format of [@easy_to_hard].
```
Comment (afspies) [inline docs/paper/paper.md:311]: we have been cited by quite a few other works, are those worth listing? I think you list some

Comment (afspies) [reply docs/paper/paper.md:311]: [Transformers Can Navigate Mazes With Multi-Step Prediction](https://arxiv.org/abs/2412.05117)
[AlphaMaze: Enhancing Large Language Models' Spatial Intelligence via GRPO](https://arxiv.org/abs/2502.14669)
[Sparkle: Mastering basic spatial capabilities in vision language models elicits generalization to composite spatial reasoning](https://arxiv.org/abs/2410.16162)
[iA: Imperative Learning-based A Search for Pathfinding](https://arxiv.org/abs/2403.15870)
[Imperative learning: A self-supervised neural-symbolic learning framework for robot autonomy](https://arxiv.org/abs/2406.16087)
T-SCEND: Test-time Scalable MCTS-enhanced Diffusion Model

**Review by afspies — COMMENTED** [review]
Great work! Left some minor comments. I assumed the choice of content matches the JOSS expectations

File: docs/paper/paper.md (hunk 6 above)
Comment (mivanit) [reply docs/paper/paper.md]: we don't actually evaluate the performance of different tokenization schemes in existing papers (although @afspies and I have been talking about doing this on pretrained LLMs), only how different maze generation algos affect the performance

I tried to elaborate on what tested means here, but it really does just mean that we dont have tests for those additional parameters

File: docs/paper/paper.md (hunk 8 above)
Comment (mivanit) [reply docs/paper/paper.md:91]: `Although mazes are well suited for these investigations, we found that existing maze generation packages [@cobbe2019procgen; @harriesMazeExplorerCustomisable3D2019; @gh_Ehsan_2022; @gh_Nemeth_2019; @easy_to_hard] lack support for transforming between multiple representations and provide limited control over the maze generation process.`

File: docs/paper/paper.md (hunk 11 above)
Comment (mivanit) [reply docs/paper/paper.md:141]: `In practice, using provided deduplication filters, we do find that exact duplicate mazes are generated very infrequently, even when generating very large datasets.`

File: docs/paper/paper.md (hunk 12 above)
Comment (mivanit) [reply docs/paper/paper.md:186]: under explained maybe, and we should also reference causal world models since we found that the semicolon tokens are so important. will need to edit this

File: docs/paper/paper.md (hunk 13 above)
Comment (mivanit) [reply docs/paper/paper.md:318]: I mention:
- [Transformers Can Navigate Mazes With Multi-Step Prediction](https://arxiv.org/abs/2412.05117)
- [iA: Imperative Learning-based A Search for Pathfinding](https://arxiv.org/abs/2403.15870)
- [Imperative learning: A self-supervised neural-symbolic learning framework for robot autonomy](https://arxiv.org/abs/2406.16087)

adding citations for:
- "T-SCEND"
- [AlphaMaze: Enhancing Large Language Models' Spatial Intelligence via GRPO](https://arxiv.org/abs/2502.14669)
